### DISCUSSION TITLE
How to crop an image and add text to it

### DISCUSSION THREAD

**Discussion by tqchiem** [ImageMagick/ImageMagick6#329] [General]
Hello,

I am using ImageMagick 6.9.11-60 on Ubuntu 22.04LTS and would like to crop an image, create a white rectangular box at the bottom (inside the image), and add black text at the center of this box. Below are the steps I do with specifics:

1). Crop an image with 445x446 dimensions:
$ convert in.jpg -crop 445x486+1327+534 out.jpg          ; (this step works)

2). Draw a white rectangular box of 445x40 at the bottom of out.jpg: 
$ convert -fill white -draw 'rectangle 0,446,445,486' out.jpg rect_out.jpg      ... (this step works)

3). Place black text at the center of the white rectangular box:
$ convert -font helvetica -fill black -pointsize 20 -gravity center -draw "text 0,476 'TEXT TO BE DISPLAYED'" rect_out.jpg Image_output.jpg   ...(this step does not place any text)

The top left hand corner of my white rectangular box is at (0,446). To vertically center my text of 20 point size, I use "text 0,476..", so I should have 10 px top and bottom margin. With this setting, no text is displayed. However, after trial and error, with the setting of "text 0, 223..", the text appears centered in the white rectangular. Any idea as to why?. 
[....the number 223 came from image height/2 minus white box height/2; or, (486/2)-(40/2)... ] 

Also, is there a more efficient way to do this without using intermediate image files, i.e., in one command or some other methods?. My application is to crop a large number of images, so fast processing time is important. 

Any help would be greatly appreciated. Thanks.

**Comment by fmw42** [comment]
Here is one way to do that.  Use parentheses processing to create the white area and use -annotate to draw the text centered in it.  Then composite that over the cropped input image.  I took some random larger image since you did not provide one.

```
magick image1.jpg -crop 445x486+1327+534 +repage \
\( -size 445x40 xc:white -draw 'rectangle 0,446,445,486' \
-font helvetica -pointsize 20 -fill black -gravity center -annotate +0+0 'TEXT TO BE DISPLAYED' \) \
-gravity south -compose over -composite \
result.jpg

```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/6061a8c3-b196-4edc-abaf-98c0d43136fc)

**Reply by tqchiem** [reply]
Thanks for your quick reply.
Yes, that is exactly what I would like to achieve as my initial step. My version only supports "convert". 
My next step is to use variables in the command so I can pass values like W, H, L, T, string using the same method as suggested above. Any suggestion would be greatly appreciated.

**Reply by fmw42** [reply]
If on IM 6, use convert in place of magick, and compare in place of magick compare. Otherwise, in this case the rest is the same.  A Windows user will need to help you with the variables.

**Comment by snibgo** [comment]
> I am using ImageMagick 6.9.11-60 on Ubuntu 22.04LTS ...

> My next step is to use variables in the command ...

I suppose you are using bash. That is a question about bash, not ImageMagick. An example of setting a variable, and using the value of the variable:
```
W=123
echo ${W}
```
The bash interpreter does the job of translating from `${W}` to `123`, and passes `123` to the `echo` command.

**Reply by fmw42** [reply]
Replace any number in the code with ${variable} for the code.  Set the variable as variable=value.

**Comment by tqchiem** [comment]
I plan to use it in Python but your suggestions work in bash, as in my test bash below:

```
#!/bin/bash

W=445
H=486
L=1327
T=534
BH=40
BT=$(($H-$BH))

#convert image1.jpg -crop 445x486+1327+534 +repage \
#\( -size 445x40 xc:white -draw 'rectangle 0,446,445,486' \
#-font helvetica -pointsize 20 -fill black -gravity center -annotate +0+0 'TEXT TO BE DISPLAYED' \) \
#-gravity south -compose over -composite \
#result.jpg

convert image1.jpg -crop ${W}x${H}+${L}+${T} +repage \
\( -size ${W}x${BH} xc:white -draw "rectangle 0,${BT},${W},${H}" \
-font helvetica -pointsize 20 -fill black -gravity center -annotate +0+0 'TEXT TO BE DISPLAYED' \) \
-gravity south -compose over -composite \
result.jpg
```

Thanks for your help. Next, I will try in Python.

**Reply by fmw42** [reply]
The text can be a variable, too. Python/OpenCV text fonts is limited.

**Comment by tqchiem** [comment]
While I read up on the options you use in the command, I think the option "+repage" is the key to solving the problem with my original command. 

It seems the software keeps track of the last reference point on the image. This option resets it to (0,0). From there, the "-gravity center" option puts the reference point back to the center of the image. Without "+repage", the reference point is elsewhere and therefore "-draw text..." would not place text where you think it should be. Also, "text 0,476..." places text off the image since the reference point is at 486/2=243. This needs to be "text 0,223..." (20 pixels off from the image bottom) to place text at the center of the rectangular box with 40 pixel height. 

I updated my original command as below and it works:

```
convert image1.jpg -crop 445x486+1327+534 \
-fill white -draw 'rectangle 0,446,445,486' +repage \
-font helvetica -pointsize 20 -fill black -gravity center -draw "text 0,223 'TEXT TO BE DISPLAYED'" result.jpg
```
Is my understanding correct of how this software works with respect to reference point and resetting it?

**Comment by snibgo** [comment]
That's not quite correct. Your "reference point" is really the "image offset on the canvas". `+repage` resets the virtual canvas.

Imagine that an IM image is printed on a piece of paper, and that paper is glued onto a blank canvas (like a painting canvas). Normally, the canvas is the same size as the paper.

When we `-crop` an image, we give a width and height, and offsets of x and y for the top-left position of the crop. These offsets are relative to the canvas, not the paper. And IM records the previous canvas size, and the offsets into that canvas.

The same applies to `-trim` and some other operations.

For example, where `toes.png` is 267x233 pixels:
```
magick toes.png -crop 50x50+50+50 +write info: -crop 20x20+60+60 info:
toes.png PNG 50x50 267x233+50+50 16-bit TrueColor sRGB 320268B 0.000u 0:00.000
toes.png PNG 20x20 267x233+60+60 16-bit TrueColor sRGB 320268B 0.000u 0:00.000
```
We crop toes.png, then crop the result. After the first crop, the image is 50x50 pixels as expected. But the virtual canvas is `267x233+50+50 `: the original size of the image, at an offset of `+50+50`.

The second crop is 20x20 pixels, at an offset of `+60+60` relative to top-left of the canvas. We couldn't use an offset of less than `+50+50` because the image isn't there on the canvas.

Now do the same, but with with `+repage` after each `-crop`:
```
magick toes.png -crop 50x50+50+50 +repage +write info: -crop 20x20+10+10 +repage info:
toes.png PNG 50x50 16-bit TrueColor sRGB 320268B 0.000u 0:00.000
toes.png PNG 20x20 16-bit TrueColor sRGB 320268B 0.000u 0:00.000
```
The first repage removes the virtual canvas metadata. In effect, the canvas is `50x50+0+0`. To get the same crop that we had above, we need offsets of `+10+10`.

Some file formats (eg PNG and MIFF) record the entire canvas metadata. Some (eg TIFF) record only offsets. Most (eg JPG) don't record any canvas metadata.

Why does IM do this? Because we might chop an image into many pieces, and do some operations on those pieces, then reassemble them. From the canvas offsets, the pieces will be put back where they came from.

**Comment by tqchiem** [comment]
Thanks for taking time to explain. I think the virtual canvas concept helps a lot with respect to offsets. I am new to IM (as of yesterday) and not quite grasp the concept of how IM works. Your explanation can be a good intro material for that. I looked into online document for draw text, until now I am not sure how it places the text even though it displays what I want. It appears to use the center of the text box for placement and not its top left corner. 

For this task, I can either go with Python/OpenCV or Python with IM via os.system(). I wonder what your thoughts are on performance and efficiency. Thanks.

**Reply by fmw42** [reply]
Python/OpenCV should be faster, but less functional for many things that IM does easily. IM is designed to handle large images (larger than your RAM).

For novices, see

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php (and see the sitemap link at the bottom)
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli
https://legacy.imagemagick.org/discourse-server/index.php
